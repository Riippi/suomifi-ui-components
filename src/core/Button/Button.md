```js
import { Button } from 'suomifi-ui-components';

<>
  <Button
    className="my-button--test"
    onClick={() => console.log('Test button click')}
    type="submit"
  >
    Button
  </Button>

  <Button disabled fullWidth>
    Button disabled fullWidth
  </Button>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button icon="login" aria-label="Login">
    Button icon="login" aria-label="Login"
  </Button>

  <Button iconRight="login" aria-labelledby="button-label">
    <span id="button-label">Login</span> iconRight="login"
    aria-labelledby="button-label"
  </Button>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <div example="negative">
    <Button.negative>Button.negative</Button.negative>

    <Button.negative disabled fullWidth icon="login">
      Button.negative disabled fullWidth icon="login"
    </Button.negative>
  </div>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button.secondary>Button.secondary</Button.secondary>

  <Button.secondary icon="login">
    Button.secondary icon="login"
  </Button.secondary>

  <Button.secondary disabled fullWidth icon="login">
    Button.secondary disabled fullWidth icon="login"
  </Button.secondary>

  <Button.secondaryNoborder>
    Button.secondaryNoborder
  </Button.secondaryNoborder>

  <Button.secondaryNoborder icon="login">
    Button icon="login"
  </Button.secondaryNoborder>

  <Button.secondaryNoborder disabled icon="login">
    Button.secondaryNoborder disabled icon="login"
  </Button.secondaryNoborder>
</>;
```

```js
import { Button } from 'suomifi-ui-components';

<>
  <Button.tertiary>Button.tertiary</Button.tertiary>

  <Button.tertiary disabled icon="login">
    Button.tertiary disabled icon="login"
  </Button.tertiary>
</>;
```
