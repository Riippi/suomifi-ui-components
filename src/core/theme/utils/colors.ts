import { ColorDesignTokens } from 'suomifi-design-tokens';
import { InternalTokensProp } from '../';

export const colorValue = (props: InternalTokensProp) => (
  colorToken: keyof ColorDesignTokens,
) => props.tokens.colors[colorToken].hsl;

export const colorsUtils = (colors: ColorDesignTokens) =>
  Object.entries(colors).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: value.hsl,
    }),
    {} as { [key in keyof ColorDesignTokens]: string },
  );
