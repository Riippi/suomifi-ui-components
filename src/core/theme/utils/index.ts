export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { spacingUtils, margin, padding, spacingModifiers } from './spacing';
export {
  typographyUtils,
  TypographyUtil,
  fontSize,
  FontSizeProps,
} from './typography';
export { colorsUtils, colorValue } from './colors';
