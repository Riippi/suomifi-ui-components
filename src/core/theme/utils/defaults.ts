import {
  suomifiTheme,
  SuomifiTheme,
  defaultTokens,
  SuomifiTokens,
  InternalTokensProp,
  PartialTokens,
} from '../';
import { asPropType, Omit } from '../../../utils/typescript';

const internalTokens = (tokens?: PartialTokens) =>
  !!tokens ? { ...defaultTokens, ...tokens } : defaultTokens;

/**
 * Check component props and do common defaulting
 * - Check if tokens are given or use default ones
 * - include internal tokens (THESE ARE NOT GEMERATED BY GIVEN TOKENS! but can be overridden by given tokens)
 * - If as-prop (styled-components) set pass it as asProp
 * @param props All component's props
 */
export const withSuomifiDefaultProps = <
  T extends { tokens?: PartialTokens; asProp?: asPropType }
>({
  tokens,
  as,
  ...props
}: T & { as?: asPropType }) =>
  ({
    ...props,
    tokens: internalTokens(tokens),
    ...(!!as ? { asProp: as } : {}),
  } as Omit<T, 'tokens'> & InternalTokensProp);

export interface Tokens {
  tokens: SuomifiTokens;
}

export interface Theme {
  theme: SuomifiTheme;
}
export type TokensOrThemeProps = Tokens | Theme;

/**
 * Return theme if defined or theme based on tokens
 * @param {Object} prop.theme Theme if defined
 * @param {Object} prop.tokens Tokens
 */
export const themeOrTokens = (props: TokensOrThemeProps) =>
  'theme' in props ? props.theme : suomifiTheme(props.tokens);
