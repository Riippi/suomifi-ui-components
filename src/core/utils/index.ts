import { FlattenSimpleInterpolation } from 'styled-components';
import { suomifiTheme, ThemeProp } from '../theme';

interface BaseStylesInterface {
  theme: ThemeProp;
  [key: string]: any;
}

/**
 *  Get CSS export from baseStyles generated by Emotion/Styled
 * @param baseStyles Emotion CSS
 * @param props properties needed to get desired styles from components
 */
export const cssFromBaseStyles = (
  baseStyles: ({ theme }: BaseStylesInterface) => FlattenSimpleInterpolation,
  props = {},
) => baseStyles({ ...props, theme: suomifiTheme }).join('');
