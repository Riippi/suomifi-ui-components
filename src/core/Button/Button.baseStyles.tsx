import { css } from '@emotion/core';
import { suomifiTheme, ThemeProp } from '../theme';
import { ButtonProps } from './Button';
import { element, focus, button } from '../theme/reset';

const fullWidthStyles = css`
  display: block;
  width: 100%;
`;

const negativeStyles = (theme: ThemeProp) => css`
  &.fi-button--negative {
    background: none;
    background-color: ${theme.colors.highlightBase};
    border: 1px solid ${theme.colors.whiteBase};
    text-shadow: none;

    &:hover {
      background: ${theme.gradients.whiteBaseNegative};
    }

    &:active {
      background: none;
      background-color: ${theme.colors.highlightBase};
    }

    &.fi-button--disabled,
    &[disabled],
    &:disabled {
      opacity: 0.5;
      background: none;
    }
  }
`;

const secondary = (theme: ThemeProp) => css`
  color: ${theme.colors.highlightBase};
  background: none;
  background-color: ${theme.colors.whiteBase};
  border: 1px solid ${theme.colors.highlightBase};
  text-shadow: none;

  &:hover {
    background: ${theme.gradients.depthLight26};
  }

  &:active {
    background: none;
    background-color: ${theme.colors.highlightLight53};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    color: ${theme.colors.depthBase};
    text-shadow: none;
    background: none;
    background-color: ${theme.colors.highlightLight53};
    border-color: ${theme.colors.depthBase};
  }
`;

const secondaryStyles = (theme: ThemeProp) => css`
  &.fi-button--secondary {
    ${secondary(theme)}
  }
`;

const secondaryNoBorderStyles = (theme: ThemeProp) => css`
  &.fi-button--secondary-noborder {
    ${secondary(theme)}
    border: none;
  }
`;

const tertiaryStyles = (theme: ThemeProp) => css`
  &.fi-button--tertiary {
    ${secondary(theme)}
    background: ${theme.colors.highlightLight50};
    border: none;

    &:hover {
      background: ${theme.colors.highlightLight53};
    }

    &:active {
      background: ${theme.colors.highlightLight53};
    }
  }
`;

export const baseStyles = ({
  theme = suomifiTheme,
  fullWidth = false,
}: ButtonProps) => css`
  ${button(theme)}
  padding: ${theme.spacing.s} ${theme.spacing.m};
  min-height: 40px;
  color: ${theme.colors.whiteBase};
  background: ${theme.gradients.highlightBase};
  border-radius: ${theme.radius.basic};
  text-align: center;
  text-shadow: ${theme.shadows.invertTextShadow};
  cursor: pointer;

  &:hover {
    background: ${theme.gradients.highlightLight4};
  }

  &:active {
    background: ${theme.gradients.highlightDark9};
  }

  &.fi-button--disabled,
  &[disabled],
  &:disabled {
    text-shadow: 0 1px 1px ${theme.colors.blackBase};
    background: ${theme.gradients.depthBase};
    pointer-events: none;
    user-select: none;
  }

  ${fullWidth && fullWidthStyles}
  ${negativeStyles(theme)}
  ${secondaryStyles(theme)}
  ${secondaryNoBorderStyles(theme)}
  ${tertiaryStyles(theme)}

  & > .fi-button_icon {
    width: 16px;
    height: 16px;
    margin-right: ${theme.spacing.s};
    vertical-align: middle;
    transform: translateY(-0.1em);
    &.fi-button_icon--right {
      margin-right: 0;
      margin-left: ${theme.spacing.s};
    }
  }
`;

export const unStyled = ({ theme = suomifiTheme }) => css`
  ${element(theme)}
  ${focus(theme)}
  border-radius: ${theme.radius.basic};
  cursor: pointer;
`;
