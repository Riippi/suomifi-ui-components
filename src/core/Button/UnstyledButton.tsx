import React from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { unStyled } from './Button.baseStyles';
import { InternalTokensProp } from '../theme';
import { Button, ButtonProps } from '../../components/Button/Button';

const baseClassName = 'fi-button';

export const UnstyledButton = styled(
  (props: ButtonProps & InternalTokensProp) => {
    const { className, ...passProps } = props;
    return (
      <Button {...passProps} className={classnames(className, baseClassName)} />
    );
  },
)`
  ${props => unStyled(props)}
`;
