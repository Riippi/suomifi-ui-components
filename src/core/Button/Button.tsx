import React, { Component } from 'react';
import styled from '@emotion/styled';
import classnames from 'classnames';
import { ThemeComponent, ThemeProp } from '../theme';
import { withDefaultTheme } from '../theme/utils';
import { baseStyles } from './Button.baseStyles';
import {
  Button as CompButton,
  ButtonProps as CompButtonProps,
} from '../../components/Button/Button';
import { Icon, IconProps, IconKeys } from '../Icon/Icon';
import { UnstyledButton } from './UnstyledButton';

type ButtonVariant =
  | 'default'
  | 'negative'
  | 'secondary'
  | 'secondary-noborder'
  | 'tertiary'
  | 'unstyled';

export interface ButtonProps extends CompButtonProps, ThemeComponent {
  /**
   * Set width to grow all available space
   */
  fullWidth?: boolean;
  /**
   * 'default' | 'negative' | 'secondary' | 'secondary-noborder' | 'tertiary'
   * @default default
   */
  variant?: ButtonVariant;
  /**
   * Icon from suomifi-theme
   */
  icon?: IconKeys;
  /**
   * Icon from suomifi-theme to be placed on right side
   */
  iconRight?: IconKeys;
  /**
   * Properties given to Icon-component
   */
  iconProps?: IconProps;
}

const baseClassName = 'fi-button';
const iconClassName = `${baseClassName}_icon`;
const iconRightClassName = `${baseClassName}_icon--right`;

const StyledButton = styled(
  ({
    theme,
    fullWidth,
    variant = 'default',
    className,
    ...passProps
  }: ButtonProps & { right?: boolean }) => (
    <CompButton
      {...passProps}
      className={classnames(className, {
        [`${baseClassName}--${variant}`]: variant !== 'default',
      })}
    />
  ),
)`
  ${props => baseStyles(props)}
`;

const iconColor = ({
  theme,
  invert,
  disabled,
}: {
  theme?: ThemeProp;
  invert?: boolean;
  disabled?: boolean;
}) => {
  if (!!theme) {
    const defaultColor = !!disabled
      ? theme.colors.depthBase
      : theme.colors.whiteBase;
    const secondaryColor = !!disabled
      ? theme.colors.depthBase
      : theme.colors.highlightBase;
    return !!invert ? secondaryColor : defaultColor;
  }
  return undefined;
};

class ButtonWithIcon extends Component<ButtonProps> {
  render() {
    const { icon, iconRight, iconProps = {}, ...passProps } = this.props;
    const { theme, disabled, variant } = passProps;
    const { className: iconPropsClassName, ...passIconProps } = iconProps;

    if (variant === 'unstyled') {
      return <UnstyledButton {...passProps} />;
    }

    const secondaryOrTertiary =
      !!variant &&
      (variant === 'secondary' ||
        variant === 'secondary-noborder' ||
        variant === 'tertiary');
    return (
      <StyledButton {...passProps}>
        {!!icon && (
          <Icon
            mousePointer={true}
            icon={icon}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            className={classnames(iconClassName, iconPropsClassName)}
            {...passIconProps}
          />
        )}
        {passProps.children}
        {!!iconRight && (
          <Icon
            mousePointer={true}
            icon={iconRight}
            color={iconColor({ theme, disabled, invert: secondaryOrTertiary })}
            className={classnames(
              iconClassName,
              iconRightClassName,
              iconPropsClassName,
            )}
            {...passIconProps}
          />
        )}
      </StyledButton>
    );
  }
}

/**
 * <i class="semantics" />
 * Use for inside Application onClick events.<br />
 * When using Button.secondaryNoborder with other than white background,<br />
 * define styles background color for all needed states (:hover, :active, :disabled)<br /><br />
 * You can use separate UnstyledButton to get Button without variant styles added by<br />
 * `import { UnstyledButton } from 'suomifi-ui-components';` or `<Button.unstyled />`.
 */
export class Button extends Component<ButtonProps> {
  static negative = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="negative" />;
  };

  static secondary = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="secondary" />;
  };

  static secondaryNoborder = (props: ButtonProps) => {
    return (
      <ButtonWithIcon
        {...withDefaultTheme(props)}
        variant="secondary-noborder"
      />
    );
  };

  static tertiary = (props: ButtonProps) => {
    return <ButtonWithIcon {...withDefaultTheme(props)} variant="tertiary" />;
  };

  static unstyled = (props: ButtonProps) => {
    return <UnstyledButton {...withDefaultTheme(props)} />;
  };

  render() {
    return <ButtonWithIcon {...withDefaultTheme(this.props)} />;
  }
}
