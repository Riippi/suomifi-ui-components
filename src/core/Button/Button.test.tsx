import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../utils/test/axe';

import { Button } from './Button';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Button.baseStyles';

const TestButton = <Button data-testid="button">Test</Button>;

test('calling render with the same component on the same container does not remount', () => {
  const buttonRendered = render(TestButton);
  const { getByTestId, container, rerender } = buttonRendered;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('button').textContent).toBe('Test');

  // re-render the same component with different props
  rerender(<Button data-testid="nottub">Test two</Button>);
  expect(getByTestId('nottub').textContent).toBe('Test two');
});

test('should not have basic accessibility issues', axeTest(TestButton));

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('color'));
});
