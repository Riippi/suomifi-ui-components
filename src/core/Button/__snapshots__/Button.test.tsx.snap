// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.emotion-0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
  color: hsl(0,0%,16%);
  font-family: 'Source Sans Pro','Helvetica Neue',Arial,sans-serif;
  font-size: 16px;
  font-weight: 600;
  line-height: 1.5;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  line-height: 1;
  padding: 8px 16px;
  min-height: 40px;
  color: hsl(0,0%,100%);
  background: linear-gradient(0deg,hsl(212,63%,45%) 0%,#4387d5 100%);
  border-radius: 2px;
  text-align: center;
  text-shadow: 0 1px 1px hsl(214,100%,24%);
  cursor: pointer;
}

.emotion-0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.emotion-0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.emotion-0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.emotion-0::-webkit-inner-spin-button {
  height: auto;
}

.emotion-0::-webkit-outer-spin-button {
  height: auto;
}

.emotion-0:focus {
  outline: 0;
  position: relative;
}

.emotion-0:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid hsl(23,82%,53%);
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 hsl(23,82%,53%);
  z-index: 9999;
}

.emotion-0:focus:not(:focus-visible):after {
  content: none;
}

.emotion-0:hover {
  background: linear-gradient(0deg,hsl(212,63%,49%) 0%,#5492d9 100%);
}

.emotion-0:active {
  background: linear-gradient(0deg,hsl(212,63%,37%) 0%,#2c72c4 100%);
}

.emotion-0.fi-button--disabled,
.emotion-0[disabled],
.emotion-0:disabled {
  text-shadow: 0 1px 1px hsl(0,0%,16%);
  background: linear-gradient(0deg,hsl(202,7%,67%) 0%,#c0c5c9 100%);
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0.fi-button--negative {
  background: none;
  background-color: hsl(212,63%,45%);
  border: 1px solid hsl(0,0%,100%);
  text-shadow: none;
}

.emotion-0.fi-button--negative:hover {
  background: linear-gradient(-180deg,rgba(255,255,255,0.1) 0%,rgba(255,255,255,0) 100%);
}

.emotion-0.fi-button--negative:active {
  background: none;
  background-color: hsl(212,63%,45%);
}

.emotion-0.fi-button--negative.fi-button--disabled,
.emotion-0.fi-button--negative[disabled],
.emotion-0.fi-button--negative:disabled {
  opacity: 0.5;
  background: none;
}

.emotion-0.fi-button--secondary {
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
}

.emotion-0.fi-button--secondary:hover {
  background: linear-gradient(0deg,hsl(202,7%,93%) 0%,hsl(0,0%,100%) 100%);
}

.emotion-0.fi-button--secondary:active {
  background: none;
  background-color: hsl(212,63%,98%);
}

.emotion-0.fi-button--secondary.fi-button--disabled,
.emotion-0.fi-button--secondary[disabled],
.emotion-0.fi-button--secondary:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.emotion-0.fi-button--secondary-noborder {
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
  border: none;
}

.emotion-0.fi-button--secondary-noborder:hover {
  background: linear-gradient(0deg,hsl(202,7%,93%) 0%,hsl(0,0%,100%) 100%);
}

.emotion-0.fi-button--secondary-noborder:active {
  background: none;
  background-color: hsl(212,63%,98%);
}

.emotion-0.fi-button--secondary-noborder.fi-button--disabled,
.emotion-0.fi-button--secondary-noborder[disabled],
.emotion-0.fi-button--secondary-noborder:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.emotion-0.fi-button--tertiary {
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
  background: hsl(212,63%,95%);
  border: none;
}

.emotion-0.fi-button--tertiary:hover {
  background: linear-gradient(0deg,hsl(202,7%,93%) 0%,hsl(0,0%,100%) 100%);
}

.emotion-0.fi-button--tertiary:active {
  background: none;
  background-color: hsl(212,63%,98%);
}

.emotion-0.fi-button--tertiary.fi-button--disabled,
.emotion-0.fi-button--tertiary[disabled],
.emotion-0.fi-button--tertiary:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.emotion-0.fi-button--tertiary:hover {
  background: hsl(212,63%,98%);
}

.emotion-0.fi-button--tertiary:active {
  background: hsl(212,63%,98%);
}

.emotion-0 > .fi-button_icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
}

.emotion-0 > .fi-button_icon.fi-button_icon--right {
  margin-right: 0;
  margin-left: 8px;
}

<button
  aria-disabled="false"
  class="fi-button emotion-0"
  data-testid="button"
  tabindex="0"
  type="button"
>
  Test
</button>
`;
